weldx/util.py

```diff
@@ -1201,3 +1193,97 @@ def time_ref(self, value: pd.Timestamp):
                 self._obj.time.attrs["time_ref"] = value  # set new time_ref value
             else:
                 self._obj.time.attrs["time_ref"] = value
+
+
+class _Eq_compare_nested:
+    """Compares nested data structures like lists, sets, tuples, arrays, etc."""
+
+    compare_funcs = {
+        (np.ndarray, pint.Quantity, pd.Index): lambda x, y: np.all(x == y),
+        (xr.DataArray, xr.Dataset): lambda x, y: x.identical(y),
+    }
+
+    @staticmethod
+    def _compare(x, y):
+        # 1. treat special cases first
+        # 2. rely upon the fact, that comparison would fail,
+        #    if y is of an incompatible type.
+        for types, func in _Eq_compare_nested.compare_funcs.items():
+            if isinstance(x, types):
+                return func(x, y)
+            return x == y
+
+        return x == y
```

Comment:
Maybe it is the missing coffee, but I don't get this. To my understanding, the loop can't execute more than once, since the body always returns. So it is obsolete. As a further consequence, the `return` after the loop should be unreachable. What am I missing?

Reply:
I was missing coffee too, I just spotted it as well. The return inside the loop made no sense.

Reply:
I guess we would have seen it in the coverage if we hadn't spotted it before 😄